### PR TITLE
[ADT] Always use 32-bit size type for SmallVector with 16-bit elements

### PR DESCRIPTION
**Pull request by jayfoad**
`SmallVector` has a special case to allow vector of char to exceed 4 GB in
size on 64-bit hosts. Apply this special case only for 8-bit element
types, instead of all element types < 32 bits.

This makes `SmallVector<MCPhysReg>` more compact because `MCPhysReg` is
`uint16_t`.

### CODE REVIEW

File: llvm/include/llvm/ADT/SmallVector.h [hunk 1]
```diff
@@ -116,8 +116,7 @@ template <class Size_T> class SmallVectorBase {
 
 template <class T>
 using SmallVectorSizeType =
-    std::conditional_t<sizeof(T) < 4 && sizeof(void *) >= 8, uint64_t,
-                       uint32_t>;
+    std::conditional_t<sizeof(T) == 1, uintptr_t, uint32_t>;
```
Comment (nikic) [inline llvm/include/llvm/ADT/SmallVector.h:119]:
```suggestion
    std::conditional_t<sizeof(T) == 1, size_t, uint32_t>;
```
Unlikely to differ, but I think that is closer to what is intended...

**Review by nikic — APPROVED** [review]
LGTM. To the best of my knowledge, the original motivation for this was specifically about the char/byte case, to support large binary blobs. If this is needed for other types, then we should have a mechanism to control it.

**Review by kuhar — COMMENTED** [review]
Please add some tests for common types like `char`, `int16_t`, etc. (e.g., using `std::is_same_v<T1, T2>`)

**Comment by jayfoad** [comment]
> Please add some tests for common types like `char`, `int16_t`, etc. (e.g., using `std::is_same_v<T1, T2>`)

There were already some static_asserts for this kind of thing. I added one for int16_t.

**Review by kuhar — APPROVED** [review]

**Comment by dyung** [comment]
> LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-darwin` running on `doug-worker-3` while building `llvm` at step 5 "build-unified-tree".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/23/builds/398
> 
> Here is the relevant piece of the build log for the reference:
> 
> ```
> Step 5 (build-unified-tree) failure: build (failure)
> ...
> 54.104 [5597/12/186] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/Record.cpp.o
> 54.498 [5596/12/187] Building CXX object utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/IntrinsicEmitter.cpp.o
> 55.096 [5595/12/188] Building CXX object lib/Option/CMakeFiles/LLVMOption.dir/Option.cpp.o
> 55.120 [5594/12/189] Building CXX object lib/Bitstream/Reader/CMakeFiles/LLVMBitstreamReader.dir/BitstreamReader.cpp.o
> 55.152 [5593/12/190] Linking CXX static library lib/libLLVMBitstreamReader.a
> 55.538 [5592/12/191] Building CXX object utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/DirectiveEmitter.cpp.o
> 55.809 [5591/12/192] Building CXX object lib/Option/CMakeFiles/LLVMOption.dir/ArgList.cpp.o
> 55.886 [5590/12/193] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/TGParser.cpp.o
> 55.924 [5589/12/194] Linking CXX static library lib/libLLVMTableGen.a
> 56.087 [5588/12/195] Linking CXX executable bin/llvm-min-tblgen
> FAILED: bin/llvm-min-tblgen 
> : && /usr/bin/clang++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names -Wl,-dead_strip -Wl,-no_warn_duplicate_libraries utils/TableGen/Basic/CMakeFiles/obj.LLVMTableGenBasic.dir/CodeGenIntrinsics.cpp.o utils/TableGen/Basic/CMakeFiles/obj.LLVMTableGenBasic.dir/SDNodeProperties.cpp.o utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/TableGen.cpp.o utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/ARMTargetDefEmitter.cpp.o utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/Attributes.cpp.o utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/DirectiveEmitter.cpp.o utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/IntrinsicEmitter.cpp.o utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/RISCVTargetDefEmitter.cpp.o utils/TableGen/CMakeFiles/llvm-min-tblgen.dir/VTEmitter.cpp.o -o bin/llvm-min-tblgen  -Wl,-rpath,@loader_path/../lib  lib/libLLVMSupport.a  lib/libLLVMTableGen.a  lib/libLLVMSupport.a  -lm  /Library/Developer/CommandLineTools/SDKs/MacOSX14.4.sdk/usr/lib/libz.tbd  /usr/local/lib/libzstd.dylib  lib/libLLVMDemangle.a && :
> Undefined symbols for architecture x86_64:
>   "llvm::SmallVectorBase<unsigned long>::grow_pod(void*, unsigned long, unsigned long)", referenced from:
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       llvm::detail::IEEEFloat::toString(llvm::SmallVectorImpl<char>&, unsigned int, unsigned int, bool) const in libLLVMSupport.a[13](APFloat.cpp.o)
>       ...
> ld: symbol(s) not found for architecture x86_64
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> 57.136 [5588/11/196] Building CXX object lib/DebugInfo/MSF/CMakeFiles/LLVMDebugInfoMSF.dir/MSFError.cpp.o
> 57.372 [5588/10/197] Building CXX object lib/DebugInfo/MSF/CMakeFiles/LLVMDebugInfoMSF.dir/MSFCommon.cpp.o
> 57.847 [5588/9/198] Building CXX object lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/CodeViewError.cpp.o
> 57.896 [5588/8/199] Building CXX object lib/DebugInfo/MSF/CMakeFiles/LLVMDebugInfoMSF.dir/MappedBlockStream.cpp.o
> 58.109 [5588/7/200] Building CXX object lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/AppendingTypeTableBuilder.cpp.o
> 58.381 [5588/6/201] Building CXX object lib/Option/CMakeFiles/LLVMOption.dir/OptTable.cpp.o
> 58.553 [5588/5/202] Building CXX object lib/DebugInfo/MSF/CMakeFiles/LLVMDebugInfoMSF.dir/MSFBuilder.cpp.o
> 58.627 [5588/4/203] Building CXX object lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/CodeViewRecordIO.cpp.o
> 59.182 [5588/3/204] Building CXX object lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/ContinuationRecordBuilder.cpp.o
> 59.308 [5588/2/205] Building CXX object lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/CVSymbolVisitor.cpp.o
> ```

This failure also happened on my other mac buildbot https://lab.llvm.org/buildbot/#/builders/190/builds/750

It didn't get automatically flagged because the host machine ran out of disk space which I am fixing up now.

**Comment by BertalanD** [comment]
In `SmallVectorSizeType`'s definition, **`size_t`** or `uint32_t` is used:

https://github.com/llvm/llvm-project/blob/2582d11f1a8a5783828156d3ced354727f422885/llvm/include/llvm/ADT/SmallVector.h#L117-L119

But the `grow_pod` function is instantiated for **`uint64_t`** or `uint32_t`:

https://github.com/llvm/llvm-project/blob/2582d11f1a8a5783828156d3ced354727f422885/llvm/lib/Support/SmallVector.cpp#L163-L170

On the arm64 Apple ABI, `size_t` == `unsigned long` and `uint64_t` == `unsigned long long`.

**Comment by chelcassanova** [comment]
This is also breaking the GreenDragon macOS buildbots: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/6522/console

**Comment by kuhar** [comment]
Revert?

**Comment by chelcassanova** [comment]
Sure, just to unblock the bots while this gets looked into.

**Pull request merged**